File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/offline/TileStoreController.kt
package com.mapbox.maps.mapbox_maps.offline

import android.content.Context
import android.os.Handler
import com.mapbox.common.TileStore
import com.mapbox.maps.OfflineManager
import com.mapbox.maps.mapbox_maps.pigeons.*
import com.mapbox.maps.mapbox_maps.toFLTTileRegion
import com.mapbox.maps.mapbox_maps.toFLTTileRegionEstimateProgress
import com.mapbox.maps.mapbox_maps.toFLTTileRegionEstimateResult
import com.mapbox.maps.mapbox_maps.toFLTTileRegionLoadProgress
import com.mapbox.maps.mapbox_maps.toFLTValue
import com.mapbox.maps.mapbox_maps.toGeometry
import com.mapbox.maps.mapbox_maps.toNetworkRestriction
import com.mapbox.maps.mapbox_maps.toResult
import com.mapbox.maps.mapbox_maps.toTileDataDomain
import com.mapbox.maps.mapbox_maps.toTileRegionEstimateOptions
import com.mapbox.maps.mapbox_maps.toTileStoreOptionsKey
import com.mapbox.maps.mapbox_maps.toTilesetDescriptorOptions
import com.mapbox.maps.mapbox_maps.toValue
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

private const val EVENT_CHANNEL_PREFIX = "com.mapbox.maps.flutter/tilestore"

class TileStoreController(
  private val context: Context,
  private val binaryMessenger: BinaryMessenger,
  private val channelSuffix: String,
  private val tileStore: TileStore
) : _TileStore {

  private val offlineManager = OfflineManager()
  private var tileRegionLoadProgressHandlers = HashMap<String, EventChannel.EventSink>()
  private var tileRegionEstimateProgressHandlers = HashMap<String, EventChannel.EventSink>()
  private val mainHandler = Handler(context.mainLooper)

  override fun loadTileRegion(
    id: String,
    loadOptions: TileRegionLoadOptions,
    callback: (Result<TileRegion>) -> Unit
  ) {
    tileStore.loadTileRegion(
      id, offlineManager.tileRegionLoadOptions(loadOptions, context),
      { progress ->
        mainHandler.post {
          tileRegionLoadProgressHandlers[id]?.success(progress.toFLTTileRegionLoadProgress().toList())
        }
      },
      { expected ->
        mainHandler.post {
          callback(expected.toResult { it.toFLTTileRegion() })
          tileRegionLoadProgressHandlers.remove(id)?.endOfStream()
        }
      }
    )
  }

  override fun addTileRegionLoadProgressListener(id: String) {
    val eventChannel = EventChannel(binaryMessenger, "com.mapbox.maps.flutter/$channelSuffix/tile-region-$id")
    eventChannel.setStreamHandler(
      object : EventChannel.StreamHandler {
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
          events?.let { tileRegionLoadProgressHandlers[id] = it }
        }
        override fun onCancel(arguments: Any?) {
          tileRegionLoadProgressHandlers.remove(id)
        }
      }
    )
  }

  override fun estimateTileRegion(
    id: String,
    loadOptions: TileRegionLoadOptions,
    estimateOptions: TileRegionEstimateOptions?,
    callback: (Result<TileRegionEstimateResult>) -> Unit
  ) {
    tileStore.estimateTileRegion(
      id,
      offlineManager.tileRegionLoadOptions(loadOptions, context),
      estimateOptions?.toTileRegionEstimateOptions() ?: com.mapbox.common.TileRegionEstimateOptions(null),
      { progress ->
        mainHandler.post {
          tileRegionEstimateProgressHandlers[id]?.success(
            progress.toFLTTileRegionEstimateProgress().toList()
          )
        }
      },
      { expected ->
        mainHandler.post {
          callback(expected.toResult { it.toFLTTileRegionEstimateResult() })
          tileRegionEstimateProgressHandlers.remove(id)?.endOfStream()
        }
      }
    )
  }

  override fun addTileRegionEstimateProgressListener(id: String) {
    val eventChannel = EventChannel(binaryMessenger, "com.mapbox.maps.flutter/$channelSuffix/tile-region-estimate$id")
    eventChannel.setStreamHandler(
      object : EventChannel.StreamHandler {
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
          events?.let { tileRegionEstimateProgressHandlers[id] = it }
        }
        override fun onCancel(arguments: Any?) {
          tileRegionEstimateProgressHandlers.remove(id)
        }
      }
    )
  }

  override fun tileRegionMetadata(id: String, callback: (Result<Map<String, Any>>) -> Unit) {
    tileStore.getTileRegionMetadata(id) { expected ->
      mainHandler.post {
        callback(expected.toResult { it.toFLTValue() as? Map<String, Any> ?: emptyMap() })
      }
    }
  }

  override fun tileRegionContainsDescriptor(id: String, options: List<TilesetDescriptorOptions>, callback: (Result<Boolean>) -> Unit) {
    val descriptors = options.map { offlineManager.createTilesetDescriptor(it.toTilesetDescriptorOptions(context)) }
    tileStore.tileRegionContainsDescriptors(id, descriptors) { expected ->
      mainHandler.post {
        callback(expected.toResult { it })
      }
    }
  }

  override fun allTileRegions(callback: (Result<List<TileRegion>>) -> Unit) {
    tileStore.getAllTileRegions { expected ->
      mainHandler.post {
        callback(expected.toResult { it.map { region -> region.toFLTTileRegion() } })
      }
    }
  }

  override fun tileRegion(id: String, callback: (Result<TileRegion>) -> Unit) {
    tileStore.getTileRegion(id) { expected ->
      mainHandler.post {
        callback(expected.toResult { it.toFLTTileRegion() })
      }
    }
  }

  override fun removeRegion(id: String, callback: (Result<TileRegion>) -> Unit) {
    tileStore.removeTileRegion(
      id,
      { expected ->
        mainHandler.post {
          callback(expected.toResult { it.toFLTTileRegion() })
        }
      }
    )
  }

  override fun setOptionForKey(key: _TileStoreOptionsKey, domain: TileDataDomain?, value: Any?) {
    domain?.also {
      tileStore.setOption(key.toTileStoreOptionsKey(), it.toTileDataDomain(), value?.toValue() ?: com.mapbox.bindgen.Value.nullValue())
    } ?: run {
      tileStore.setOption(key.toTileStoreOptionsKey(), value?.toValue() ?: com.mapbox.bindgen.Value.nullValue())
    }
  }
}

private fun OfflineManager.tileRegionLoadOptions(fltValue: TileRegionLoadOptions, context: Context): com.mapbox.common.TileRegionLoadOptions {
  val builder = com.mapbox.common.TileRegionLoadOptions.Builder()
    .geometry(fltValue.geometry?.toGeometry())
    .metadata(fltValue.metadata?.toValue())
    .acceptExpired(fltValue.acceptExpired)
    .networkRestriction(fltValue.networkRestriction.toNetworkRestriction())
    .startLocation(fltValue.startLocation)
    .averageBytesPerSecond(fltValue.averageBytesPerSecond?.toInt())
    .extraOptions(fltValue.extraOptions?.toValue())

  fltValue.descriptorsOptions?.let { options ->
    val descriptors: List<TilesetDescriptorOptions> = options.filterNotNull()
    builder.descriptors(
      descriptors.map { createTilesetDescriptor(it.toTilesetDescriptorOptions(context)) }
    )
  }

  return builder.build()
}

File: ios/Classes/Offline/TileStoreController.swift
import MapboxMaps
import Flutter

final class TileStoreController: _TileStore {

    enum `Error`: Swift.Error {
        case invalidTileRegionLoadOptions
    }

    private let messenger: SuffixBinaryMessenger
    private let tileStore: TileStore

    private lazy var offlineManager = OfflineManager()
    private var tileRegionLoadProgressHandlers: [String: AnyFlutterStreamHandler] = [:]
    private var tileRegionEstimateProgressHandlers: [String: AnyFlutterStreamHandler] = [:]

    init(messenger: SuffixBinaryMessenger, tileStore: TileStore) {
        self.messenger = messenger
        self.tileStore = tileStore
    }

    func loadTileRegion(id: String, loadOptions: TileRegionLoadOptions, completion: @escaping (Result<TileRegion, Swift.Error>) -> Void) {
        guard let loadOptions = offlineManager.tileRegionLoadOptions(loadOptions) else {
            completion(.failure(Error.invalidTileRegionLoadOptions))
            return
        }

        tileStore.loadTileRegion(forId: id, loadOptions: loadOptions) { [weak self] progress in
            guard let self else { return }
            self.tileRegionLoadProgressHandlers[id]?.eventSink?(progress.toFLTTileRegionLoadProgress().toList())
        } completion: { [weak self] result in
            executeOnMainThread(completion)(result.map { $0.toFLTTileRegion() })
            self?.tileRegionLoadProgressHandlers.removeValue(forKey: id)
        }
    }

    func addTileRegionLoadProgressListener(id: String) throws {
        let handler = AnyFlutterStreamHandler()
        let eventChannel = FlutterEventChannel(name: "com.mapbox.maps.flutter/\(messenger.suffix)/tile-region-\(id)", binaryMessenger: messenger.messenger)
        eventChannel.setStreamHandler(handler)
        tileRegionLoadProgressHandlers[id] = handler
    }

    func estimateTileRegion(id: String, loadOptions: TileRegionLoadOptions, estimateOptions: TileRegionEstimateOptions?, completion: @escaping (Result<TileRegionEstimateResult, Swift.Error>) -> Void) {

        guard let loadOptions = offlineManager.tileRegionLoadOptions(loadOptions) else {
            completion(.failure(Error.invalidTileRegionLoadOptions))
            return
        }

        tileStore.estimateTileRegion(
            forId: id,
            loadOptions: loadOptions,
            estimateOptions: estimateOptions.map(MapboxCommon.TileRegionEstimateOptions.init(fltValue:))) { [weak self] progress in
                guard let self else { return }
                self.tileRegionEstimateProgressHandlers[id]?.eventSink?(progress.toFLTTileRegionEstimateProgress().toList())
            } completion: { [weak self] result in
                executeOnMainThread(completion)(result.map { $0.toFLTTileRegionEstimateResult() })
                self?.tileRegionEstimateProgressHandlers.removeValue(forKey: id)
            }
    }

    func addTileRegionEstimateProgressListener(id: String) throws {
        let handler = AnyFlutterStreamHandler()
        let eventChannel = FlutterEventChannel(name: "com.mapbox.maps.flutter/\(messenger.suffix)/tile-region-estimate-\(id)", binaryMessenger: messenger.messenger)
        eventChannel.setStreamHandler(handler)
        tileRegionEstimateProgressHandlers[id] = handler
    }

    func tileRegionMetadata(id: String, completion: @escaping (Result<[String: Any], Swift.Error>) -> Void) {
        tileStore.tileRegionMetadata(forId: id) { result in
            completion(result.map { $0 as? [String: Any] ?? [:] })
        }
    }

    func tileRegionContainsDescriptor(id: String, options: [TilesetDescriptorOptions], completion: @escaping (Result<Bool, Swift.Error>) -> Void) {
        let descriptors = options
            .compactMap(MapboxCoreMaps.TilesetDescriptorOptions.init(fltValue:))
            .map(offlineManager.createTilesetDescriptor(for:))

        tileStore.tileRegionContainsDescriptors(forId: id, descriptors: descriptors, completion: executeOnMainThread(completion))
    }

    func allTileRegions(completion: @escaping (Result<[TileRegion], Swift.Error>) -> Void) {
        tileStore.allTileRegions { result in
            let result = result.map { regions in
                regions.map { $0.toFLTTileRegion() }
            }
            completion(result)
        }
    }

    func tileRegion(id: String, completion: @escaping (Result<TileRegion, Swift.Error>) -> Void) {
        tileStore.tileRegion(forId: id) { result in
            executeOnMainThread(completion)(result.map { $0.toFLTTileRegion() })
        }
    }

    func removeRegion(id: String, completion: @escaping (Result<TileRegion, Swift.Error>) -> Void) {
        tileStore.removeRegion(forId: id) { result in
            executeOnMainThread(completion)(result.map { $0.toFLTTileRegion() })
        }
    }

    func setOptionForKey(key: _TileStoreOptionsKey, domain: TileDataDomain?, value: Any?) throws {
        if let domain {
            tileStore.setOptionForKey(key.toTileStoreOptionsKey(), domain: domain.toTileDataDomain(), value: value as Any)
        } else {
            tileStore.setOptionForKey(key.toTileStoreOptionsKey(), value: value as Any)
        }
    }
}

extension OfflineManager {

    fileprivate func tileRegionLoadOptions(_ fltValue: TileRegionLoadOptions) -> MapboxCommon.TileRegionLoadOptions? {
        .init(
            geometry: convertDictionaryToGeometry(dict: fltValue.geometry),
            descriptors: fltValue.descriptorsOptions?.compactMap { descriptorOptions in
                guard let descriptorOptions else { return nil }
                return MapboxCoreMaps.TilesetDescriptorOptions(fltValue: descriptorOptions).map(createTilesetDescriptor(for:))
            },
            metadata: fltValue.metadata,
            acceptExpired: fltValue.acceptExpired,
            networkRestriction: MapboxCommon.NetworkRestriction(fltValue: fltValue.networkRestriction),
            averageBytesPerSecond: fltValue.averageBytesPerSecond.map(Int.init),
            extraOptions: fltValue.extraOptions
        )
    }
}
